github-actions/feature-request/src/octokit.ts

```diff
@@ -0,0 +1,259 @@
+/**
+ * The following file contains a modified version of the Microsoft implementation of
+ * communication with GitHub's REST APIs.
+ *
+ * The original is available at
+ * https://github.com/microsoft/vscode-github-triage-actions/blob/eb561150d9bfab77954cfda7ffef149c07e0e079/api/octokit.ts
+ */
+
+import { GitHub } from '@actions/github';
+import { Octokit } from '@octokit/rest';
+import { Comment, GitHubAPI, GitHubIssueAPI, Issue, Query } from './api';
+import { log } from './log';
+
+export class OctoKit implements GitHubAPI {
+  private _octokit: GitHub;
+
+  // The organization members will likely not change
+  // between issues. We want to cache them so we
+  // don't query the GitHub API for each issue.
+  private _orgMembers = new Set<string>();
+
+  protected get octokit(): GitHub {
+    return this._octokit;
+  }
+
+  protected mockLabels: Set<string> = new Set();
+
+  constructor(
+    protected token: string,
+    protected params: { repo: string; owner: string },
+    protected options: { readonly: boolean } = { readonly: false },
+  ) {
+    this._octokit = new GitHub(token);
+  }
+
+  async *query(query: Query): AsyncIterableIterator<GitHubIssueAPI> {
+    const q = `${query.q} repo:${this.params.owner}/${this.params.repo}`;
+
+    const options = this.octokit.search.issuesAndPullRequests.endpoint.merge({
+      ...query,
+      q,
+      per_page: 100,
+      // To access reactions we need a specific media type we specify via the accept header
+      // https://docs.github.com/en/rest/reference/repos#list-commit-comments-for-a-repository-preview-notices
+      headers: { Accept: 'application/vnd.github.squirrel-girl-preview+json' },
+    });
+
+    let pageNum = 0;
+
+    const timeout = async () => {
+      if (pageNum < 2) {
+        /* pass */
+      } else if (pageNum < 4) {
+        await new Promise((resolve) => setTimeout(resolve, 10000));
+      } else {
+        await new Promise((resolve) => setTimeout(resolve, 30000));
+      }
+    };
+
+    for await (const pageResponse of this.octokit.paginate.iterator(options)) {
+      await timeout();
+      const page: Array<Octokit.SearchIssuesAndPullRequestsResponseItemsItem> = pageResponse.data;
+      log(`Page ${++pageNum}: ${page.map(({ number }) => number).join(' ')}`);
+      for (const issue of page) {
+        yield new OctoKitIssue(
+          this.token,
+          this.params,
+          this.octokitIssueToIssue(issue),
+          this.options,
+        );
+      }
+    }
```

Comment:
What do you think of doing this instead and eliminating the iterator entirely?

This just has github paginate and then merge the pages together, allowing you to perform the transformation just before merging into a single list.  The return type of this is `Promise<OctoKitIssue[]>`, which I think ends up being easier to work with.  Additionally, it removes the need for the backoff timeouts as Github handles it for us.


```ts
    return await this.octokit.paginate(
      options,
      (page: Octokit.Response<Octokit.SearchIssuesAndPullRequestsResponseItemsItem[]>) => {
        return page.data.map(issue => new OctoKitIssue(
          this.token,
          this.params,
          this.octokitIssueToIssue(issue),
          this.options,
          ));
        });
      }
```

Reply:
I see we've already used this in the issue lock bot, so seems like `paginate` will simplify the implementation!

Reply:
I haven't seen us running into out of memory issues, though I didn't think about that being an issue.  I am fine with either way you want to go about it.